### ISSUE TITLE
Stats Reporting Stereo for Multichannel Bitstream

### ISSUE THREAD

**Issue by Doofussy2** [mpv-player/mpv#6334]
**mpv version and platform**

Shinchiro build Nov 18, but it's with all builds

**Reproduction steps**

Enable audio bitstreaming and play any 5.1 or 6.1 audio. 7.1 appears to report, correctly. The audio is correctly bitstreamed. 

**Expected behavior**

Correctly report how many channels there are.

**Actual behavior**

mpv stats always reports stereo
 
**Log file**

[Portable mpv log.txt](https://github.com/mpv-player/mpv/files/2611876/Portable.mpv.log.txt)

**Comment by Doofussy2** [comment]
This is also the same with the recent fix to play DTS-MA HRA. And I've read a report of this on RPi, also.

**Comment by Argon-** [comment]
The stats script doesn't come up with these values, it just asks mpv for the channel count which then seems to report an incorrect value. 

Can you try something? Download the script locally and remove the following lines:
https://github.com/mpv-player/mpv/blob/master/player/lua/stats.lua#L531 531, 533, 535
Run mpv with `--load-stats-overlay=no` to disable the internalversion

**Comment by tomty89** [comment]
Isn't it normal that it reports stereo unless it's a TrueHD or HDMA stream?

`[ 0.568][v][af] [in] 48000Hz stereo 2ch spdif-dts` matches with:
https://github.com/mpv-player/mpv/blob/master/audio/decode/ad_spdif.c#L226

**Comment by Doofussy2** [comment]
@Argon- 

> Edit: your file is detected as stero:
[ 0.568][v][af] [in] 48000Hz stereo 2ch spdif-dts

You can see from that same log, that the audio is 5.1
`[   0.023][i][cplayer]  (+) Audio --aid=1 --alang=eng (*) 'DTS-ES' (dts 6ch 48000Hz)`

It's as @tomty89 says. mpv is labeling it as stereo.

This is what I think should be cahnged. As is, it's implying that mpv is downmixing to 2 channels before bitstreaming. And the stats are showing it as stereo, when it isn't. It's misleading. Preferably, it should report the actual channel count. Or at the very least, remove 'stereo 2ch'.

**Comment by tomty89** [comment]
AFAIK while spdif output is often known as "bitstreaming"/"pass-through", the audio track is not exactly sent to the receiver without being touched/transformed at all (tho it's not a lossy process). That's why there is #5703.

While the "stat" does not provide you with the actual spec of the track, it does inform you how it is presented to the audio device (or in other word, how the audio device is being opened). So it's not really "misleading".

I am not sure how easy or difficult it would be to make that "stat" show the spec of the track, but AFAICT, structurally speaking it would be an ugly thing to do. Neither do I find it necessary, as to some extent that's the job of the receiver/decoder. Not to mention that mpv is already providing us with the info, as you quoted. It is also reporting something real (and useful), as I mentioned.

**Comment by Doofussy2** [comment]
@tomty89 
Ok, I can understand your point. 

@Argon- 
Is it possible to have the stats report spdif-ac3, or spdif-dts etc, and not 'stereo'? Or just say "Audio Passthrough"?

**Comment by Argon-** [comment]
@Doofussy2 If there is a property carrying this information then we can use it.
Currently the stats display the following audio related properties:

* `audio-codec`: https://mpv.io/manual/master/#command-interface-audio-codec
* `audio-params/format`: https://mpv.io/manual/master/#command-interface-audio-params/format
* `audio-params/samplerate`: https://mpv.io/manual/master/#command-interface-audio-params/samplerate
* `audio-params/channel-count`: https://mpv.io/manual/master/#command-interface-audio-params/channel-count
* `packet-audio-bitrate`: https://mpv.io/manual/master/#command-interface-packet-sub-bitrate
* `af`: https://mpv.io/manual/master/#command-interface-af

Apparently the wanted information is not among them and `audio-params/channel-count` is not reporting the correct/desired value.

We are not using `audio-codec-name` so far as I assume(d) it's only a subset of `audio-codec`. I can test this when you have me a spdif-dts/whatever sample file. The only other audio related property not shown is the channel layout.

**Comment by Doofussy2** [comment]
Thanks @Argon-. It's more about not saying it's stereo, than having the correct channel count. Having it say stereo, is implying down-mixing. Even if when bitstreaming the channel count were blank, that would be an improvement over misinformation. Ideally, it would give the correct channel count. But if not possible, leaving it blank when bitstreaming the audio, would be better. As for a sample, you can use any 5.1 AC3, DTS, EAC3 etc. They all present this way, when bitstreaming. Disabling passthrough, the channel count is correct.

**Comment by Argon-** [comment]
You've got me an example?

**Comment by Doofussy2** [comment]
I can share something via a google drive, but you guys don't want that. I haven't figured out how to upload to the suggested location. How would you like me to share it?

**Comment by Argon-** [comment]
I just tested a couple files but as @tomty89 said, the channel count is correct for TrueHD. So just bitstreaming is not necessarily a problem, it depends on the exact codec being used whether the channel-count property will be useful or "wrong".
The easy way out would be to disable the channel-count when a user set the `audio-spdif` option.

**Comment by Doofussy2** [comment]
> The easy way out would be to disable the channel-count when a user set the audio-spdif option.

That may be a good way, because I've just noticed that bitstreaming 5.1 DTS-HD,  the logs/stats report it as 7.1.

`[   0.026][i][cplayer]  (+) Audio --aid=1 --alang=eng (*) (dts 6ch 48000Hz)`
`[   0.656][v][af] [in] 192000Hz 7.1 8ch spdif-dtshd`

So it's misleading on both counts

**Comment by Doofussy2** [comment]
And it's also the same for 5.1 TrueHD.

You can get samples, here
https://thedigitaltheater.com/dolby-trailers/

**Comment by tomty89** [comment]
There's no need/point to "test", as I told you whenever you do "spdif" output, the stream *has to be* presented in a certain hard coded spec to the audio interface according to their codec type. I even gave you a link to the relevant code.

**Comment by Argon-** [comment]
Well augmented as in "hide certain fields in case `audio-spdif` was set", if that's desirable. 
The correct values you seek are unfortunately not available as properties.

**Comment by Doofussy2** [comment]
Well, as I said, using a generic term like bitstreaming or leaving it blank, I think would be better than having it report wrong information.

**Comment by tliron** [comment]
I know this issue was closed a long time ago, but I spent hours trying to debug what was going on due to the stats showing "Channels: 2" even though I was clearly hearing more than 5.1 channels in SPDIF passthrough ("bitstream":).

I understand the logic of showing the number of channels actually going to the decoder, but perhaps we could compromise by also showing stats about the actual encoding of the audio track? For conciseness we can only show them if there is a difference between their values and what is being sent to the decoder.
